plugin/trino-elasticsearch/src/main/java/io/trino/plugin/elasticsearch/ElasticsearchMetadata.java

```diff
@@ -579,16 +585,87 @@ public ConnectorTableProperties getTableProperties(ConnectorSession session, Con
         return Optional.of(new ConstraintApplicationResult<>(handle, TupleDomain.withColumnDomains(unsupported), newExpression, false));
     }
 
-    protected static String likeToRegexp(String like)
+    protected static boolean checkArgument(Call call, int index, Predicate<ConnectorExpression> predicate)
+    {
+        return predicate.test(call.getArguments().get(index));
+    }
+
+    protected static boolean isLikeCall(Call call)
+    {
+        if (!LIKE_PATTERN_FUNCTION_NAME.equals(call.getFunctionName())) {
+            return false;
+        }
+
+        if (!checkArgument(call, 0, Variable.class::isInstance)
+            && checkArgument(call, 1, Constant.class::isInstance)) {
+            return false;
+        }
+
+        if (call.getArguments().size() == 3) {
+            return checkArgument(call, 2, Constant.class::isInstance);
+        }
+
+        return true;
+    }
+
+    protected static String likeToRegexp(Slice pattern, Optional<Slice> escape)
     {
-        // TODO: This can be done more efficiently by using a state machine and iterating over characters (See io.trino.type.LikeFunctions.likePattern(String, char, boolean))
-        String regexp = like.replaceAll(Pattern.quote("\\"), Matcher.quoteReplacement("\\\\")); // first, escape regexp's escape character
-        for (char c : REGEXP_RESERVED_CHARACTERS) {
-            regexp = regexp.replaceAll(Pattern.quote(String.valueOf(c)), Matcher.quoteReplacement("\\" + c));
-        }
-        return regexp
-                .replaceAll("%", ".*")
-                .replaceAll("_", ".");
+        String patternString = pattern.toStringUtf8();
+        char escapeChar = escape.map(ElasticsearchMetadata::getEscapeChar)
+                .orElse('0');
+        StringBuilder regex = new StringBuilder();
+        boolean escaped = false;
+        for (char currentChar : patternString.toCharArray()) {
```

Comment:
`toCharArray` makes a copy of the data (already second time)

you can iterate over codepoints in Slice directly.
see https://github.com/trinodb/trino/blob/ddea70ad46eef4f835dc887036f4a96bec3b2976/core/trino-main/src/main/java/io/trino/type/LikeFunctions.java#L125-L126 how to do this

when working with codepoints you can use `StringBuilder.appendCodePoint` method